src/target/riscv/riscv.c

```diff
@@ -1636,59 +2070,56 @@ static int riscv_hit_watchpoint(struct target *target, struct watchpoint **hit_w
 	}
 
 	riscv_reg_t dpc;
-	riscv_get_register(target, &dpc, GDB_REGNO_DPC);
+	if (riscv_get_register(target, &dpc, GDB_REGNO_DPC) != ERROR_OK)
+		return ERROR_FAIL;
 	const uint8_t length = 4;
 	LOG_TARGET_DEBUG(target, "dpc is 0x%" PRIx64, dpc);
 
 	/* fetch the instruction at dpc */
 	uint8_t buffer[length];
 	if (target_read_buffer(target, dpc, length, buffer) != ERROR_OK) {
-		LOG_TARGET_ERROR(target, "Failed to read instruction at dpc 0x%" PRIx64, dpc);
+		LOG_TARGET_ERROR(target, "Failed to read instruction at dpc 0x%" PRIx64,
+						 dpc);
 		return ERROR_FAIL;
 	}
 
-	uint32_t instruction = 0;
+	riscv_insn_t instruction = 0;
 
 	for (int i = 0; i < length; i++) {
 		LOG_TARGET_DEBUG(target, "Next byte is %x", buffer[i]);
 		instruction += (buffer[i] << 8 * i);
 	}
 	LOG_TARGET_DEBUG(target, "Full instruction is %x", instruction);
 
-	/* find out which memory address is accessed by the instruction at dpc */
-	/* opcode is first 7 bits of the instruction */
-	uint8_t opcode = instruction & 0x7F;
-	uint32_t rs1;
-	int16_t imm;
-	riscv_reg_t mem_addr;
+	int rs;
+	target_addr_t mem_addr;
+	int16_t memoffset;
+
+	if (get_loadstore_membase_regno(target, instruction, &rs) != ERROR_OK)
+		return ERROR_FAIL;
+	if (riscv_get_register(target, &mem_addr, rs) != ERROR_OK)
+		return ERROR_FAIL;
+	if (get_loadstore_memoffset(target, instruction, &memoffset) != ERROR_OK)
+		return ERROR_FAIL;
 
-	if (opcode == MATCH_LB || opcode == MATCH_SB) {
-		rs1 = (instruction & 0xf8000) >> 15;
-		riscv_get_register(target, &mem_addr, rs1);
+	mem_addr += memoffset;
+	bool is_load;
 
-		if (opcode == MATCH_SB) {
-			LOG_TARGET_DEBUG(target, "%x is store instruction", instruction);
-			imm = ((instruction & 0xf80) >> 7) | ((instruction & 0xfe000000) >> 20);
-		} else {
-			LOG_TARGET_DEBUG(target, "%x is load instruction", instruction);
-			imm = (instruction & 0xfff00000) >> 20;
-		}
-		/* sign extend 12-bit imm to 16-bits */
-		if (imm & (1 << 11))
-			imm |= 0xf000;
-		mem_addr += imm;
-		LOG_TARGET_DEBUG(target, "Memory address=0x%" PRIx64, mem_addr);
-	} else {
-		LOG_TARGET_DEBUG(target, "%x is not a RV32I load or store", instruction);
+	if (verify_loadstore(target, instruction, &is_load) != ERROR_OK)
 		return ERROR_FAIL;
-	}
 
 	struct watchpoint *wp = target->watchpoints;
 	while (wp) {
-		/*TODO support length/mask */
-		if (wp->address == mem_addr) {
+		/* TODO support mask and check read/write/access */
+		if (mem_addr >= wp->address &&
+			mem_addr < (wp->address + wp->length)) {
```

Comment:
This condition is not quite correct. RISCV specification recommends to perform multi-value matches. That is we should take the size of underlying memory reference into account. Maybe this should be a target parameter.

Reply:
Where can I read about this?

Reply:
https://github.com/riscv/riscv-debug-spec/releases/download/1.0.0-rc2/riscv-debug-specification.pdf 

Search for the following lines:

```
This bit determines the contents of the XLEN-bit compare
values.
0 (address): There is at least one compare value and it
contains the lowest virtual address of the access. It is
recommended that there are additional compare values for
the other accessed virtual addresses. (E.g. on a 32-bit read
from 0x4000, the lowest address is 0x4000 and the other
addresses are 0x4001, 0x4002, and 0x4003.)
```

Reply:
Thanks for the clarification.
I think that to check this condition it will be necessary to add the setting of the "Select" bit when programming the trigger (as described in Debugger Implementation for Triggers). At the moment, I don't see any use of the "Select" bit in the OpenOCD code when programming the Triggers. 
Maybe another MR is needed to solve this issue?

Reply:
> he "Select" bit when programming the trigger

Setting select bit is about data watchpoints. These are different. Current implementation does not set it because it always sets address watchpoints.

> Maybe another MR is needed to solve this issue?

Well, I'm fine with that - would you add a TODO then if it's not a bit dial?

Reply:
> Well, I'm fine with that - would you add a TODO then if it's not a bit dial?

Ok.
Thank you.

Reply:
I agree with @aap-sc that there should be a check for intersection of the access range and watchpoint range. 

And I also agree that it is perfectly fine with me to just add a `TODO` comment for later resolution. This PR is already complex enough :)

Reply:
Added.
Thank you!